vaas/vaas/vcl/renderer.py

```diff
@@ -186,23 +187,50 @@ def __init__(self, varnish, input_data):
             'mesh_routing': varnish.cluster.service_mesh_routing
         }
 
+    def filter_cluster_domain_mappings(
+        self, cluster: LogicalCluster
+    ) -> dict[str, list[str]]:
+        """Returns a dictionary where keys are domains, and values are mapped to by the key"""
+        # we can PROBABLY assume that each domain has only one mapping
+        # in an experiment trying to add a new mapping for a source domain removed the old one
+        destination_dict = {}
+        for domain_mapping in self.input.domain_mappings:
+            domain = domain_mapping.domain
+            destination_dict[domain] = domain_mapping.mapped_domains(cluster)
+        return destination_dict
+
+    def provide_related_domains(self, cluster: LogicalCluster) -> list[str]:
+        """Returns a list of all domains related to a cluster - for static mappings,
+        the list is specified within the mapping, for dynamic - clusters are associated using
+        labels (mapping targets can contain labels, and each cluster can have multiple labels, but
+        each label belongs to exactly one cluster)"""
+        result = {m.domain for m in self.input.static_domain_mappings_by_cluster[cluster.name]}
+        for m in self.input.mapping_provider.mappings["dynamic"]:
+            if m.is_cluster_related_by_labels(cluster):
+                result.add(m.domain)
+        return sorted(list(result))
+
     @collect_processing
     def prepare_redirects(self) -> Dict[str, List[VclRedirect]]:
         redirects = {}
-        related_domains = MappingProvider(DomainMapping.objects.all()).provide_related_domains(self.varnish.cluster)
+        related_domains = self.provide_related_domains(self.varnish.cluster)
+        destinations_dict = self.filter_cluster_domain_mappings(self.varnish.cluster)
         for redirect in self.input.redirects:
```

Comment:
The better way will be an iteration on the prepared redirects for related domains.
I have prepared a draft PR how it can look like:
https://github.com/allegro/vaas/pull/519